### PR TITLE
Initial version of game 2048

### PR DESCRIPTION
**Pull request by dmydlo**

### CODE REVIEW

**Review by mateuszdargacz — COMMENTED** [review]
Bootstrap jednak nie z npm?

File: src/components/game/layout.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+interface IProps {
+  game: any;
+  dispatch: Function;
+}
+
+class GameLayout extends React.Component<IProps, {}> {
+  public componentDidMount() {
+    this.props.dispatch(start());
+    document.addEventListener('keyup', this.determineMove);
```
Comment (PythonicNinja) [inline src/components/game/layout.tsx]: You should attach event and remove it from document once it's unmounted. Here is reference of component lifecycle: https://www.tutorialspoint.com/reactjs/reactjs_component_life_cycle.htm

File: src/components/game/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  public render() {
+    return (
+      <div>
+        <header>
+          <ReactRouter.Link to="/" ><button className="btn btn-primary" >Home</button></ReactRouter.Link>
+          <button className="btn btn-warning" onClick={this.props.dispatch.bind(this, (start()))}>Restart</button>
+          <button className="btn btn-success disabled score" >{this.props.game.score}</button>
+        </header>
+        <div className="game_container">
+          {this.props.game.board.map((row: any, rowIndex: number) => {
```
Comment (PythonicNinja) [inline src/components/game/layout.tsx]: remove types of any and provide some types

File: src/components/game/layout.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+              return (<div className={fieldClass}>{field}</div>);
+            })}</div>);
+          })}
+        </div>
+      </div>
+    );
+  }
+  private determineMove = (e: any) => {
+    if (this.props.game.gameOver) {
+      return 0;
```
Comment (PythonicNinja) [inline src/components/game/layout.tsx]: maybe `this.props.dispatch(gameOver());`

File: src/components/game/layout.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+const mapStateToProps = (state: any) => {
+  return state;
+};
+
+const actions: any = { start };
+
+const dispatchToProps = (dispatch: any) => {
+  const object: any = Object;
```
Comment (PythonicNinja) [inline src/components/game/layout.tsx]: that was my hack to make it work, but there must be cleaner way

File: src/components/layout.scss [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    -moz-transition:    $transition;
+    -o-transition:      $transition;
+    -webkit-transition: $transition;
+    transition:         $transition;
+}
+
+.container {
+  margin: 0;
+  padding: 0;
+  width: 100vw;
```
Comment (PythonicNinja) [inline src/components/layout.scss]: use `100%` and add:
```
body { margin: 0 }
```
then they will work in very same way.

File: src/components/layout.scss [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  height: 100vh;
```
Comment (PythonicNinja) [inline src/components/layout.scss]: same issue here use `100%` and add `body { margin: 0 }`

File: src/components/layout.scss [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  .content {
+    margin: auto;
+    width: 100vh;
+    max-width: 100%;
+
+    .game_container {
+      margin: auto;
+      background: #bbada0;
+
+      .v0 { color: #776e65; background: #cdc1b4; }
```
Comment (PythonicNinja) [inline src/components/layout.scss]: colors into sass variables

File: src/components/layout.scss [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  .content {
+    margin: auto;
+    width: 100vh;
+    max-width: 100%;
+
+    .game_container {
+      margin: auto;
+      background: #bbada0;
```
Comment (PythonicNinja) [inline src/components/layout.scss]: color into sass variable

File: src/reducers/game.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        points += line[newIndex - 1];
+        changed = true;
+      } else if (newIndex !== index) {
+        line[newIndex] = line[index];
+        line[index] = 0;
+        changed = true;
+      }
+    }
+  }
+  return [changed, line, points];
```
Comment (PythonicNinja) [inline src/reducers/game.ts]: return Object and make interface out of it.

signature of that function would be:
`const compressLine = (line: ILine): ICompresedLine => {`

where:
```
interface ICompresedLine {
  changed: boolean;
  line:  ILine;
  points: number;
}
```

File: src/reducers/game.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+};
+
+const makeMoveForDirection = (game: any, direction: any): [any, number] => {
```
Comment (PythonicNinja) [inline src/reducers/game.ts]: convert game into object and make inteface of it

File: src/reducers/game.ts [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  let points = 0;
+  let size = game[0].length;
```
Comment (PythonicNinja) [inline src/reducers/game.ts]: in that case `game.board_size` would be nicer

File: src/reducers/game.ts (hunk 10 above)
Comment (PythonicNinja) [inline src/reducers/game.ts]: make interface out of returned object

**Pull request merged**